tests/unit/model/lifecycle-callbacks-test.js

```diff
@@ -1,299 +1,1118 @@
+
+    assert.strictEqual(
+      lifecycleEventMethodCalls,
+      0,
+      'We do not trigger didDelete when we unload a record'
+    );
   });
 
-  const Adapter = DS.Adapter.extend({
-    findRecord(store, type, id, snapshot) {
-      return { data: { id: 1, type: 'person', attributes: { name: 'Foo' } } };
-    },
+  test('becameInvalid() triggers when an update rejects with an error for a member', async function(assert) {
+    let lifecycleEventMethodCalls = 0;
+    class Person extends Model {
+      @attr
+      name;
 
-    updateRecord(store, type, snapshot) {
-      assert.equal(
-        callCount,
-        0,
-        'becameInvalid callback was not called until recordWasInvalid is called'
-      );
+      becameInvalid() {
+        lifecycleEventMethodCalls++;
+        assert.equal(this.get('isSaving'), false, 'record should not be saving');
+        assert.equal(this.get('hasDirtyAttributes'), true, 'record should not be dirty');
+      }
+    }
+    this.owner.register('model:person', Person);
+    this.owner.register('serializer:application', JSONAPISerializer);
 
-      return reject(
-        new DS.InvalidError([
+    class AppAdapter extends Adapter {
+      updateRecord() {
+        assert.equal(
+          lifecycleEventMethodCalls,
+          0,
+          'becameInvalid callback was not called before adapter.updateRecord resolves'
+        );
+        let error = new InvalidError([
           {
             title: 'Invalid Attribute',
             detail: 'error',
             source: {
-              pointer: '/data/attributes/bar',
+              pointer: '/data/attributes/name',
             },
           },
-        ])
-      );
-    },
-  });
+        ]);
 
-  let store = createStore({
-    adapter: Adapter,
-    person: Person,
-  });
+        return reject(error);
+      }
+    }
+    this.owner.register('adapter:application', AppAdapter);
+
+    let person = store.push({
+      data: {
+        type: 'person',
+        id: '1',
+        attributes: {
+          name: 'Great Name!',
+        },
+      },
+    });
+    assert.strictEqual(
+      lifecycleEventMethodCalls,
+      0,
+      'We do not trigger becameInvalid when we first push the record to the store'
+    );
+    lifecycleEventMethodCalls = 0;
 
-  let asyncPerson = run(() => store.findRecord('person', 1));
-  assert.equal(callCount, 0, 'precond - becameInvalid callback was not called yet');
+    person.set('name', 'Bad Bad Name');
 
-  // Make sure that the error handler has a chance to attach before
-  // save fails.
-  return run(() => {
-    return asyncPerson.then(person => {
-      return run(() => {
-        person.set('bar', 'Bar');
-        return person.save().catch(reason => {
-          assert.ok(reason.isAdapterError, 'reason should have been an adapter error');
+    await person.save().catch(reason => {
+      assert.strictEqual(lifecycleEventMethodCalls, 1, 'We trigger becameInvalid when we reject');
```

Comment:
Just curious, what are the reasons for preferring `strictEqual` over `equal` in some cases?

Reply:
I use `strictEqual` when we do expect strict referential equality. The advantage is two-fold: (1) it avoids a set of bugs that happen when qunit attempts to do deep recursive "looks like the same thing" comparisons when strict equality is not matched (on records this is a mega perf hit in addition to being risky), and it is an improvement over the alternative pattern of `assert.ok(foo === bar, '')` wherein if the assertion fails there is no additional information printed about what the incorrect value is without stepping into the test.